src/systems/filecoin_mining/storage_mining/storage_miner_actor.go

```diff
@@ -243,6 +302,80 @@ func (a *StorageMinerActorCode_I) _expirePreCommittedSectors(rt Runtime) {
 
 }
 
+func (st *StorageMinerActorState_I) _updateSectorUtilization(rt Runtime) []deal.DealID {
+	// TODO: verify if we should update Sector utilization for failing sectors
+	// depends on decision around collateral requirement for inactive power
+	// and what happens when a failing sector expires
+
+	// this may not work
+	ret := deal.CompactDealSet(make([]byte, 0))
+
+	for _, sectorNo := range st.Impl().ProvingSet_.SectorsOn() {
+		utilizationInfo, found := st.SectorUtilization()[sectorNo]
+
+		if !found {
+			rt.Abort("sm._updateSectorUtilization: sector not found in SectorUtilization")
+		}
+
+		newUtilization := utilizationInfo.CurrUtilization()
+
+		currEpoch := rt.CurrEpoch()
+		firstExpirationEpoch := utilizationInfo.DealExpirationQueue().Peek().Expiration()
+		totalDealCount := len(st.Sectors()[sectorNo].SealCommitment().DealIDs())
+		newExpiredDealIDs := deal.CompactDealSet(make([]byte, totalDealCount))
+
+		if firstExpirationEpoch < currEpoch {
+			// this deal has expired
+			newExpiredDeal := utilizationInfo.DealExpirationQueue().Pop()
+			newUtilization -= newExpiredDeal.PayloadPower()
+			st.SectorUtilization()[sectorNo].Impl().ActiveDealIDs_.Remove(newExpiredDeal.DealID())
+			newExpiredDealIDs.Add(newExpiredDeal.DealID())
+
+			// this is a go while loop
```

Comment:
this is weird, and wrong, what are you trying to do?

Reply:
go through all expiring deals and settle payments. trying to piggyback off the heavy lifting in updateSectorUtilization

Reply:
i think this works:
```go
queue := utilizationInfo.DealExpirationQueue()
for queue.Peek().Expiration() < currEpoch {
    expiredDeal := queue.Pop()
    newUtilization -= expiredDeal.PayloadPower()

    // state changes
    st.SectorUtilization()[sectorNo].Impl().ActiveDealIDs_.Remove(newExpiredDeal.DealID())
    newExpiredDealIDs.Add(newExpiredDeal.DealID())
}
```